README.md
# TP_Reseaux_1

history.txt
Member 2 : jdnfds

Member 1 : fgd

Member 3 : fgdgdf

Member 1 : dfgd

Member 2 : fdgd



Client.java
/***
 * EchoClient
 * Example of a TCP client 
 * Date: 10/01/04
 * Authors:
 */


import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.*;
import java.net.*;

import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;



public class Client {
	

	
  /**
  *  main method
  *  accepts a connection, receives a message from client then sends an echo to the client
  *  
  **/
	
    public static void main(String[] args) throws IOException {

        Socket echoSocket = null;
        PrintStream socOut = null;
        BufferedReader stdIn = null;
        BufferedReader socIn = null;

        // args doit prendre comme arg l'adresse IP et le port 
        // add IP : localhost
        // port > 1024 pour etre un serveur
        if (args.length != 2) {
          System.out.println("Usage: java EchoClient <EchoServer host> <EchoServer port>");
          System.exit(1);
        }

        try {
      	    // creates the socket ==> connection
      	    echoSocket = new Socket(args[0],new Integer(args[1]).intValue());
		          	    
      	  InterfaceClient ic = new InterfaceClient(echoSocket);
          ic.frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
          ic.frame.setVisible(true);
      	    
      	    //reading
			ClientThread ctSendingToServer = new ClientThread(echoSocket,true,ic);
			ctSendingToServer.start();
			
			//listening
			ClientThread ctReceivingFromServer = new ClientThread(echoSocket,false,ic);
			ctReceivingFromServer.start();
			
			//getting the history
			File file = new File("history.txt");
			if (!file.exists()) 
			{		
				file.createNewFile();
			}
			
			FileInputStream load = new FileInputStream(file);
			
			BufferedReader reader = new BufferedReader(new InputStreamReader(load));
	        String line;
	        while ((line = reader.readLine()) != null) {
	        	//writes the history in the chat window
	        	ic.messageArea.append(line + "\n");
	        	System.out.println(line);
	        }
	        reader.close();
			load.close();
			
			while(true)
			{

				if(!ctSendingToServer.keepOn) break;
			}
			
        } catch (UnknownHostException e) {
            System.err.println("Don't know about host:" + args[0]);
            System.exit(1);
        } catch (IOException e) {
            System.err.println("Couldn't get I/O for "
                               + "the connection to:"+ args[0]);
            System.exit(1);
        }
                             
        String line;
      
      System.out.println("Program terminated.");
    }
}




Server.java
/***
 * EchoServer
 * Example of a TCP server
 * Date: 10/01/04
 * Authors:
 */


import java.io.*;
import java.net.*;
import java.util.*;

public class Server  {
 	/**
  	* main method
	* @param EchoServer port
  	* 
  	**/
    public static void main(String args[]){ 
        ServerSocket listenSocket;
        BufferedReader stdIn = null;
        PrintStream socOut = null;
        BufferedReader socIn = null;
  
  	if (args.length != 1) {
          System.out.println("Usage: java EchoServer <EchoServer port>");
          System.exit(1);
  	}
	try {

		/**
		 * connection to server
		 */
		listenSocket = new ServerSocket(Integer.parseInt(args[0])); //port
		System.out.println("Server ready..."); 
		
		while (true) {
			Socket clientSocket = listenSocket.accept();
			System.out.println("Connexion from:" + clientSocket.getInetAddress());
			//reading thread
			ServerThread ctSendingToClient = new ServerThread(clientSocket,true);
			ctSendingToClient.start();
			
			//listening thread
			ServerThread ctReceivingFromClient = new ServerThread(clientSocket,false);
			ctReceivingFromClient.start();
			
			//print the history of messages in the chat
			for(int i=0; i<ServerThread.history.size(); i++) 
			{
				ctSendingToClient.runSending(ServerThread.history.get(i));
			}
	
		}
        } catch (Exception e) {
            System.err.println("Error in Server:" + e);
        }
      }
  }


ClientThread.java
/***
 * ClientThread
 * Example of a TCP server
 * Date: 14/12/08
 * Authors:
 */

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.*;
import java.net.*;
import java.util.LinkedList;

import javax.swing.JFrame;
import javax.swing.JOptionPane;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;

public class ClientThread extends Thread {
	
	private Socket clientSocket;
	private InterfaceClient ic;
	
	private boolean sender;
	public boolean keepOn = true;

	ClientThread(Socket s, boolean send, InterfaceClient ic) {
		this.clientSocket = s;
		this.sender = send;
		this.ic = ic;
	}
	
	/**
	 * receives a request from client then sends an echo to the client
	 * 
	 * @param clientSocket
	 *            the client socket
	 **/

	public void run() {

		if (sender) {
			runSending();
		} else {
			runReceiving();
		}
	}

	public void runSending() {

		try {
			

		} catch (Exception e) {
			System.err.println("Error in runSending:" + e);
		}
	}

	public void runReceiving() {
		try {
			System.out.println("received");
			BufferedReader socIn = null;
			socIn = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));

			while (true) {
				String out = socIn.readLine(); // lit depuis le socket
				//writes received messages in the chat window
				ic.messageArea.append(out + "\n");
				System.out.println(out);
			}

		} catch (Exception e) {
			System.err.println("Error in runReceiving:" + e);
		}

	}

}


InterfaceClient.java
import java.awt.Color;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.PrintStream;
import java.io.PrintWriter;
import java.net.Socket;

import javax.swing.JFrame;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;
import javax.swing.JTextField;

public class InterfaceClient {

	BufferedReader in;
	PrintStream out;
	JFrame frame = new JFrame("Chat");
	JTextField textField = new JTextField(40);
	JTextArea messageArea = new JTextArea(40, 40);
	
	/**
	 * interface constructor
	 * @param clientSocket 
	 * @throws IOException
	 */

	public InterfaceClient(Socket clientSocket) throws IOException {
		
		out = new PrintStream(clientSocket.getOutputStream());
		textField.setEditable(true);
		textField.setSelectedTextColor(Color.BLUE);
		messageArea.setEditable(false);
		frame.getContentPane().add(textField, "North");
		frame.getContentPane().add(new JScrollPane(messageArea), "Center");
		frame.pack();

		// Add Listeners
		textField.addActionListener(new ActionListener() {
			/**
			 * @param e
			 *  sends the message when the user touches "ENTER"
			 */
			public void actionPerformed(ActionEvent e) {
				// writes the message on the socket 
				out.println(textField.getText());
				textField.setText("");
			}
		});
	}

}


ServerThread.java

/***
 * ServerThread
 * Example of a TCP server
 * Date: 14/12/08
 * Authors:
 */

import java.io.*;
import java.net.*;
import java.util.LinkedList;

/**
 * 
 *
 */
public class ServerThread extends Thread {
	static LinkedList<ServerThread> listThreadsSending = new LinkedList<>();
	static LinkedList<ServerThread> listThreadsReceiving = new LinkedList<>();
	static LinkedList<String> history = new LinkedList<>(); //history of messages

	private Socket clientSocket;
	private int id;
	private boolean sender;
	public boolean keepOn = true;

	ServerThread(Socket s, boolean send) {
		this.clientSocket = s;
		this.sender = send;
		if (send) {
			listThreadsSending.add(this);
			id = listThreadsSending.size();
		} else {
			listThreadsReceiving.add(this);
			id = listThreadsReceiving.size();
		}
	}

	/**
	 * depending on sender value (true/false), choose the right method
	 */
	public void run() {

		if (sender) {
			runSending(null);
		} else {
			runReceiving();
		}
	}

	/**
	 * write in the socket
	 * 
	 * @param msg: message to write in the socket
	 */

	public void runSending(String msg) {

		try {
			PrintStream socOut = new PrintStream(clientSocket.getOutputStream());
			if (msg != null)
				socOut.println(msg);

		} catch (Exception e) {
			System.err.println("Error in runSending:" + e);
		}
	}

	/**
	 * read from the socket 
	 */
	public void runReceiving() {
	try {
		BufferedReader socIn = null;
		socIn = new BufferedReader(new InputStreamReader(clientSocket.getInputStream()));
	
			while (true) {
				String out = socIn.readLine(); // read from the socket

				// id attribution to each memeber of the conversation
				out = "Member " + this.id + " : " + out;

				// send the read information to all members connected to the chat
				for (int i = 0; (i < listThreadsSending.size()); i++) {
					listThreadsSending.get(i).runSending(out);
				}

				// creating a history file with all messages
				FileWriter fw = new FileWriter("history.txt",true);
				BufferedWriter bw = new BufferedWriter(fw);
				bw.write(out+"\n");
				bw.close();
			} 
		}catch (Exception e) {
				System.err.println("Error in runReceiving:" + e);
			}

		}

	}
